src/effection/events.ts

```diff
@@ -15,11 +15,19 @@ export function on(emitter: EventEmitter, eventName: EventName): Controller {
   }
 }
 
-export function watch(emitter: EventEmitter, names: EventName | EventName[]): Controller {
+function defaultPrepareMessage(...args) {
+  return { event: this.event, args: args };
+}
+
+export function watch(
+  emitter: EventEmitter,
+  names: EventName | EventName[],
+  prepare: (...args: any[]) => any = defaultPrepareMessage // eslint-disable-line @typescript-eslint/no-explicit-any
```

Comment:
TypeScript is not happy about the lack of types on this function, but I can't think of any sensible thing to type this as. Do you have any suggestions?

Reply:
You can always parameterize the function:

```ts
export function watch<MessageIn, MessageOut>(
  emitter: EventEmitter,
  names: EventName | EventName[],
  prepare: (...args: MessageIn[]) => MessageOut = defaultPrepareMessage // eslint-disable-line @typescript-eslint/no-explicit-any
): Controller {
  return (execution) => {
    for(let name of [].concat(names)) {
      let listener = (...args) => {
        execution.send(prepare.apply({ event: name }, args));
      }

      emitter.on(name, listener);
      execution.atExit(() => {
        emitter.off(name, listener);
      });
    }
    execution.resume(emitter);
  }
}
```